kafka/kafka.go

```diff
@@ -247,85 +244,68 @@ func (cm *consumerMessage) DiscardPayload() {
 	cm.cm = nil
 }
 
-func (ams *asyncMessageSource) ConsumeMessages(ctx context.Context, messages chan<- substrate.Message, acks <-chan substrate.Message) error {
+type consumerGroupHandler struct {
+	ctx context.Context
 
-	c, err := cluster.NewConsumerFromClient(ams.client, ams.consumerGroup, []string{ams.topic})
-	if err != nil {
-		return err
-	}
-
-	defer func() {
-		_ = c.Close()
-	}()
-
-	toAck := make(chan *consumerMessage)
-	eg, ctx := errgroup.WithContext(ctx)
-	eg.Go(func() error {
-		return ams.passMessagesToClient(ctx, c.Messages(), messages, toAck)
-	})
-	eg.Go(func() error {
-		return ams.passAcksToKafka(ctx, c, acks, toAck)
-	})
-
-	return eg.Wait()
+	messages chan<- substrate.Message
+	acks     <-chan substrate.Message
 }
 
-func (ams *asyncMessageSource) passMessagesToClient(ctx context.Context, fromKafka <-chan *sarama.ConsumerMessage, messages chan<- substrate.Message, toAck chan<- *consumerMessage) error {
+// Setup is run at the beginning of a new session, before ConsumeClaim.
+func (_ *consumerGroupHandler) Setup(_ sarama.ConsumerGroupSession) error { return nil }
+
+// Cleanup is run at the end of a session, once all ConsumeClaim goroutines have exited
+// but before the offsets are committed for the very last time.
+func (c *consumerGroupHandler) Cleanup(_ sarama.ConsumerGroupSession) error { return nil }
+
+// ConsumeClaim must start a consumer loop of ConsumerGroupClaim's Messages().
+// Once the Messages() channel is closed, the Handler must finish its processing
+// loop and exit.
+func (c *consumerGroupHandler) ConsumeClaim(sess sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	// This function can be called concurrently for multiple claims, so the code
+	// below, absent locking etc may seem wrong, but it's actually fine.
+	// Different partition claims can be processed concurrently, but we funnel
+	// them all into c.messages.  The caller puts all acks into c.acks and it
+	// doesn't matter which one of us processes the offset marking because they
+	// all work with the same session.
 	for {
 		select {
-		case <-ctx.Done():
-			return nil
-		case msg := <-fromKafka:
-			message := &consumerMessage{cm: msg}
-
-			select {
-			case <-ctx.Done():
+		case <-c.ctx.Done():
+			return c.ctx.Err()
+		case m, ok := <-claim.Messages():
+			if !ok {
 				return nil
-			case toAck <- message:
 			}
 			select {
-			case <-ctx.Done():
-				return nil
-			case messages <- message:
+			case c.messages <- &consumerMessage{cm: m}:
+			case <-c.ctx.Done():
+				return c.ctx.Err()
+			}
+		case a := <-c.acks:
+			cm := a.(*consumerMessage)
```

Comment:
I think there is still value in keeping the ack ordering check here, as otherwise consumers might accidentally acknowledge messages they haven't processed yet...

Reply:
> @mjgarton Discussing on Platform standup this morning, that we want to move forward with testing - I did explain it was unfinished.
> 
> So ignore that I moved it from draft -> review.

I suspect the first thing is that `go test` in `kafka` directory should pass.  (Note that this has never been fixed up to run in CI, so until that happens, it's a local only thing.)